libminifi/src/utils/Environment.cpp

```diff
@@ -0,0 +1,192 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utils/Environment.h"
+
+#ifdef WIN32
+#include <Windows.h>
+#else
+#include <cstdlib>
+#include <cerrno>
+#include <unistd.h>
+#endif
+#include <mutex>
+#include <vector>
+#include <iostream>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+bool Environment::runningAsService_(false);
+
+void Environment::accessEnvironment(const std::function<void(void)>& func) {
+  static std::recursive_mutex environmentMutex;
+  std::lock_guard<std::recursive_mutex> lock(environmentMutex);
+  func();
+}
+
+std::pair<bool, std::string> Environment::getEnvironmentVariable(const char* name) {
+  bool exists = false;
+  std::string value;
+
+  Environment::accessEnvironment([&exists, &value, name](){
+#ifdef WIN32
+    std::vector<char> buffer(32767U);  // https://docs.microsoft.com/en-gb/windows/win32/api/processenv/nf-processenv-getenvironmentvariablea
+    // GetEnvironmentVariableA does not set last error to 0 on success, so an error from a pervious API call would influence the GetLastError() later,
+    // so we set the last error to 0 before calling
+    SetLastError(ERROR_SUCCESS);
+    uint32_t ret = GetEnvironmentVariableA(name, buffer.data(), buffer.size());
```

Comment:
nit - can be used GetEnvironmentVariable.

Reply:
I have been bitten before by a rouge header appearing that included `Windows.h` with `UNICODE` defined, making the macros use the W versions of functions, causing segfaults when called with `char*` strings, so I prefer to explicitly define either the ANSI or the Unicode version of the winapi functions. I think it improves readability as well.